service-worker.js

```diff
@@ -59,54 +77,151 @@ var regexpExcludedURLSchema = /^(?:chrome-extension|example-extension):/i;
+            return Promise.all(
+                requests.map(function (request) {
+                    return fetch(request).then(function (response) {
+                        // Fail on 404, 500 etc
+                        if (!response.ok) throw Error('Could not fetch ' + request.url);
+                        return cache.put(request.url.replace(/\?v[^?/]+$/, ''), response);
+                    }).catch(function (err) {
+                        console.error('There was an error pre-caching files', err);
+                    });
+                })
+            );
+        })
+    );
 });
 
+// Allow sw to control current page
 self.addEventListener('activate', function (event) {
-    // "Claiming" the ServiceWorker is necessary to make it work right away,
-    // without the need to reload the page.
-    // See https://developer.mozilla.org/en-US/docs/Web/API/Clients/claim
-    event.waitUntil(self.clients.claim());
+    console.log('[SW] Claiming clients for current page');
+    // Check all the cache keys, and delete any old caches
+    event.waitUntil(
+        caches.keys().then(function (keyList) {
+            return Promise.all(keyList.map(function (key) {
+                console.log('[SW] Current cache key is ' + key);
+                if (key !== APP_CACHE && key !== ASSETS_CACHE) {
+                    console.log('[SW] App updated to version ' + appVersion + ': deleting old cache')
+                    return caches.delete(key);
+                }
+            }));
+        })
+    );
 });
 
 var outgoingMessagePort = null;
 var fetchCaptureEnabled = false;
 
 self.addEventListener('fetch', function (event) {
-    if (fetchCaptureEnabled &&
-        regexpZIMUrlWithNamespace.test(event.request.url) &&
-        event.request.method === "GET") {
-
-        // The ServiceWorker will handle this request either from CACHE_NAME or from app.js
-
-        event.respondWith(
-            // First see if the content is in the cache
-            fromCache(event.request).then(
-                function (response) {
-                    // The response was found in the cache so we respond with it 
+    // Only cache GET requests
+    if (event.request.method !== "GET") return;
+    // Remove any querystring before requesting from the cache
+    var rqUrl = event.request.url.replace(/\?[^?]+$/i, '');
+    // Select cache depending on request format
+    var cache = /\.zim\w{0,2}\//i.test(rqUrl) ? ASSETS_CACHE : APP_CACHE;
+    event.respondWith(
+        // First see if the content is in the cache
+        fromCache(cache, rqUrl).then(function (response) {
+            // The response was found in the cache so we respond with it 
+            return response;
+        }, function () {
+            // The response was not found in the cache so we look for it in the ZIM
+            // and add it to the cache if it is an asset type (css or js)
+            if (/\.zim\w{0,2}\//i.test(rqUrl) && regexpZIMUrlWithNamespace.test(rqUrl)) {
```

Comment:
If we change to prepending the ZIM name as defined in the ZIM's metadata, we should change this regex (as it won't need to account for split files).

Reply:
I have changed the corresponding function in app.js so that it uses the zimFile property `name` that we define when the archive is first loaded. Therefore the filename will never be `.zimaa`, etc., any more, so we do not need to test for two letters after the .zim extension. There actually is no ZIM metadata (in the M namespace) that corresponds to the exact archive version. It would be possible to construct one from `name` and `date`, which might be preferable to relying on the abstracted filename? See https://wiki.openzim.org/wiki/Metadata

Reply:
I think that the standardized ZIM file name is enough

Reply:
OK